### PR TITLE
Log when requested TTL isnt granted for a cert

### LINKED ISSUE #8845: Be more clear about why a requested session TTL is not granted.
### What
What would you like Teleport to do differently?

When running `tctl auth sign` to generate an identity file, there should be immediate feedback if the requested TTL is not honored.

### How
How would you implement this?

For example, if my "terraform" user has a "plugins" role with `max_session_ttl: 8h` and I do `tctl auth sign -o terraform.identity --format file --user terraform --ttl 8766h`, it should mention why:

```
WARN: requested ttl of 8766h cannot be granted. User terraform has role "plugins" with max_session_ttl 8h. proceeding with 8h ttl.
The credentials have been written to terraform.identity
```

Or if a user who has authenticated via tsh with a lower than allowed ttl tries to generate credentials for themselves that last longer:

```
WARN requested ttl of 8766h cannot be granted. Existing session has a TTL that ends in 7h56m. proceeding with 7h56m ttl.
The credentials have been written to terraform.identity
```

### Why

There are many reasons that the requested maximum ttl might not be granted. This leads to confusion and delays-- especially among teleport cloud users who must use all remote tctl commands to generate the identity files. It isn't always obvious as to why the desired results aren't happening.

### Workaround
This is a quality of life improvement to get folks on the right track in-product.

### PR DESCRIPTION
**Pull request by lxea**
When a requested TTL is not being used when calling `tctl auth sign` as it is longer than the max_session_ttl specified in a users roles the following is logged:

```
WARN             requested ttl of 4000h0m0s was not granted. user may have a role with a shorter max session ttl or an existing session ending before the requested TTL. proceeding with 4h59m58s common/auth_command.go:655
```

Fixes #8845

### CODE REVIEW

File: tool/tctl/common/auth_command.go [hunk 1]
```diff
@@ -748,6 +754,53 @@ func (a *AuthCommand) checkProxyAddr(clusterAPI auth.ClientI) error {
...
+			)
+		}
+		return nil
+	}
+	user, err := clusterAPI.GetUser(a.genUser, true)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, userRole := range user.GetRoles() {
+		role, err := clusterAPI.GetRole(context.TODO(), userRole)
```
Comment (codingllama) [inline tool/tctl/common/auth_command.go:788]: Ditch context.TODO and pass the context as a parameter. I realize it won't go much far, but it's still better context discipline.

File: tool/tctl/common/auth_command.go [hunk 2]
```diff
@@ -748,6 +754,53 @@ func (a *AuthCommand) checkProxyAddr(clusterAPI auth.ClientI) error {
 	return trace.BadParameter("couldn't find registered public proxies, specify --proxy when using --format=%q", identityfile.FormatKubernetes)
 }
 
+func (a *AuthCommand) logUngrantedTTL(clusterAPI auth.ClientI) error {
+	rc := clockwork.NewRealClock()
```
Comment (codingllama) [inline tool/tctl/common/auth_command.go:758]: A local-variable real clock doesn't have much of a reason for existing, ditch for time.Now()?

File: tool/tctl/common/auth_command.go [hunk 3]
```diff
@@ -748,6 +754,53 @@ func (a *AuthCommand) checkProxyAddr(clusterAPI auth.ClientI) error {
...
+		remaining := status.ValidUntil.Sub(rc.Now())
+		if a.genTTL > remaining {
+			logrus.Warnf("requested ttl of %s cannot be granted. "+
+				"Existing session has a TTL that ends in %s. proceeding with %s ttl.",
+				a.genTTL, remaining, remaining,
+			)
+		}
+		return nil
+	}
+	user, err := clusterAPI.GetUser(a.genUser, true)
```
Comment (codingllama) [inline tool/tctl/common/auth_command.go:783]: This is a lot of client-side logic for a few log messages, including things I assume we already check during cert generation (parsing proxy address, checking statuses, etc). Could we simplify it?

A few ideas:
* if we print the messages _after_ the certs are generated, could we drop some of the checks?
* if we wait for cert generation, could we check if the user got the requested TTLs (and hence skip all this)?
* could we get away with a simpler message and still address the issue (say, dropping the need to read roles)?

Something else that strikes me is that, if any of this fails, we are failing cert generation. That seems a bit off.

Comment (lxea) [reply tool/tctl/common/auth_command.go:783]: > if we wait for cert generation, could we check if the user got the requested TTLs (and hence skip all this)?

This seems like the most sensible option to me, however (I'm fairly sure I'm missing something here)  when I get the valid time for the cert from `key.TeleportTLSCertValidBefore` its just using the requested ttl, despite my user having a max time ttl of 5 hours in all of their roles.

Calling tctl auth sign like this:
```
 tctl sign --format file --user admin --ttl 400h -o admin.identify
```
The ttl doesnt seem to be getting truncated to 5hours acording to the contents of the cert, it also doesnt get truncated after I `tsh login`.

Comment (codingllama) [reply tool/tctl/common/auth_command.go:783]: Huh, interesting. I assumed that would be the case, but apparently I'm wrong. :)

File: tool/tctl/common/auth_command_test.go [hunk 4]
```diff
@@ -71,6 +71,14 @@ func TestAuthSignKubeconfig(t *testing.T) {
 		SigningAlg: types.CertAuthoritySpecV2_RSA_SHA2_512,
 	})
 	require.NoError(t, err)
+	user, err := types.NewUser("testuser")
```
Comment (codingllama) [inline tool/tctl/common/auth_command_test.go:74]: nit: vertical space to separate blocks?

```suggestion

	user, err := types.NewUser("testuser")
```

File: tool/tctl/common/auth_command_test.go [hunk 5]
```diff
@@ -512,6 +542,14 @@ func TestGenerateAppCertificates(t *testing.T) {
 			ClusterName: clusterNameStr,
 		})
 	require.NoError(t, err)
+	user, err := types.NewUser("testuser")
```
Comment (codingllama) [inline tool/tctl/common/auth_command_test.go:545]: nit: visually separate blocks

```suggestion

	user, err := types.NewUser("testuser")
```

**Review by codingllama — COMMENTED** [review]
Adding a couple comments on tests.

File: tool/tctl/common/auth_command.go [hunk 6]
```diff
@@ -748,6 +754,53 @@ func (a *AuthCommand) checkProxyAddr(clusterAPI auth.ClientI) error {
...
+	var addr string
+	var u *utils.NetAddr
+	if a.proxyAddr != "" {
+		var err error
+		u, err = utils.ParseAddr(a.proxyAddr)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		addr = u.Addr
+	}
```
Comment (espadolini) [inline tool/tctl/common/auth_command.go:768]: nit: `u` can stay scoped inside the block below
```suggestion
	if a.proxyAddr != "" {
		u, err := utils.ParseAddr(a.proxyAddr)
		if err != nil {
			return trace.Wrap(err)
		}
		addr = u.Addr
	}
```

File: tool/tctl/common/auth_command.go (hunk 3 above)
Comment (lxea) [reply tool/tctl/common/auth_command.go:783]: I had actually set teleport up incorrectly I think so, that method works, I've updated it with an error message  that is a bit less specific but now doesnt require getting users/roles etc.

File: tool/tctl/common/auth_command.go [hunk 7]
```diff
@@ -644,6 +644,20 @@ func (a *AuthCommand) generateUserKeys(clusterAPI auth.ClientI) error {
 		return trace.Wrap(err)
 	}
 	fmt.Printf("\nThe credentials have been written to %s\n", strings.Join(filesWritten, ", "))
+
+	expires, err := key.TeleportTLSCertValidBefore()
+	if err != nil {
+		logrus.Warnf("Failed to check TTL validity: %v.", trace.Wrap(err))
+		return nil
```
Comment (codingllama) [inline tool/tctl/common/auth_command.go:651]: A couple of suggestions here:

1. (nit) Alias the "logrus" import to "log"

2. Log using WithError, don't fail:

```suggestion
		logrus.WithError(err).Warn("Failed to check TTL validity")
		// err swallowed on purpose
```

File: tool/tctl/common/auth_command.go [hunk 8]
```diff
@@ -644,6 +644,20 @@ func (a *AuthCommand) generateUserKeys(clusterAPI auth.ClientI) error {
...
+	}
+	certRequestTime := time.Now()
+	if certRequestTime.Add(a.genTTL).Sub(expires) > time.Minute {
+		logrus.Warnf("requested ttl of %s was not granted. user may have a role with a shorter max session ttl"+
+			" or an existing session ending before the requested TTL. proceeding with %s",
+			a.genTTL,
+			expires.Sub(certRequestTime).Round(time.Second))
```
Comment (codingllama) [inline tool/tctl/common/auth_command.go:658]: nit: Capitalize messages? It's fine to do for log statements.

```suggestion
		logrus.Warnf("Requested TTL of %s was not granted. User may have a role with a shorter max session TTL"+
			" or an existing session ending before the requested TTL. Proceeding with %s.",
			a.genTTL,
			expires.Sub(certRequestTime).Round(time.Second))
```

**Review by codingllama — COMMENTED** [review]
Looks great, thanks for the fixes.

File: tool/tctl/common/auth_command.go [hunk 9]
```diff
@@ -644,6 +644,20 @@ func (a *AuthCommand) generateUserKeys(clusterAPI auth.ClientI) error {
...
 	}
 	fmt.Printf("\nThe credentials have been written to %s\n", strings.Join(filesWritten, ", "))
+
+	expires, err := key.TeleportTLSCertValidBefore()
+	if err != nil {
+		logrus.Warnf("Failed to check TTL validity: %v.", trace.Wrap(err))
+		return nil
+	}
+	certRequestTime := time.Now()
+	if certRequestTime.Add(a.genTTL).Sub(expires) > time.Minute {
```
Comment (espadolini) [inline tool/tctl/common/auth_command.go:654]: Couldn't we compare `expires` against the `Expires` of the `proto.UserCertsRequest` from above instead?

File: tool/tctl/common/auth_command.go (hunk 8 above)
Comment (espadolini) [reply tool/tctl/common/auth_command.go:658]: Shouldn't we output these messages to stdout or stderr? The user at the console might want to know that there's some discrepancy between the requested and the granted TTLs.

**Review by espadolini — APPROVED** [review]

**Review by codingllama — APPROVED** [review]

**Pull request merged**